File: android/multando-sdk/src/main/kotlin/com/multando/sdk/services/InfractionService.kt
package com.multando.sdk.services

import com.multando.sdk.core.HttpClient
import com.multando.sdk.models.InfractionResponse

/**
 * Fetches and caches infraction types.
 */
class InfractionService internal constructor(
    private val httpClient: HttpClient
) {
    private var cache: List<InfractionResponse>? = null
    private var cacheTimestamp: Long = 0
    private val cacheTtlMs: Long = 300_000 // 5 minutes

    /** List all available infraction types. Results are cached for 5 minutes. */
    suspend fun list(forceRefresh: Boolean = false): List<InfractionResponse> {
        if (!forceRefresh) {
            cache?.let {
                if (System.currentTimeMillis() - cacheTimestamp < cacheTtlMs) return it
            }
        }

        val infractions = httpClient.request<List<InfractionResponse>>(
            method = "GET",
            path = "/api/v1/infractions"
        )

        synchronized(this) {
            cache = infractions
            cacheTimestamp = System.currentTimeMillis()
        }

        return infractions
    }

    /** Fetch a single infraction by ID. */
    suspend fun get(id: String): InfractionResponse =
        httpClient.request(method = "GET", path = "/api/v1/infractions/$id")

    /** Clear the in-memory cache. */
    fun clearCache() {
        synchronized(this) {
            cache = null
            cacheTimestamp = 0
        }
    }
}

File: ios/Sources/MultandoSDK/Services/InfractionService.swift
import Foundation

/// Fetches and caches infraction types.
public final class InfractionService: @unchecked Sendable {

    private let httpClient: HTTPClient
    private var cache: [InfractionResponse]?
    private var cacheTimestamp: Date?
    private let cacheTTL: TimeInterval = 300 // 5 minutes
    private let lock = NSLock()

    init(httpClient: HTTPClient) {
        self.httpClient = httpClient
    }

    /// List all available infraction types. Results are cached for 5 minutes.
    public func list(forceRefresh: Bool = false) async throws -> [InfractionResponse] {
        if !forceRefresh, let cached = cachedValue() {
            return cached
        }

        let infractions: [InfractionResponse] = try await httpClient.request(
            method: "GET",
            path: "/api/v1/infractions"
        )

        lock.lock()
        cache = infractions
        cacheTimestamp = Date()
        lock.unlock()

        return infractions
    }

    /// Fetch a single infraction by ID.
    public func get(id: String) async throws -> InfractionResponse {
        try await httpClient.request(
            method: "GET",
            path: "/api/v1/infractions/\(id)"
        )
    }

    /// Clear the in-memory cache.
    public func clearCache() {
        lock.lock()
        cache = nil
        cacheTimestamp = nil
        lock.unlock()
    }

    private func cachedValue() -> [InfractionResponse]? {
        lock.lock()
        defer { lock.unlock() }
        guard let cache, let ts = cacheTimestamp, Date().timeIntervalSince(ts) < cacheTTL else {
            return nil
        }
        return cache
    }
}
